### ISSUE TITLE
Add "debug" command

### ISSUE THREAD

**Issue by matz-e** [matz-e/lobster#304]
It would be really helpful to have a `debug` command that runs a task reproducible and standalone… in a temporary directory (`/tmp/lobster-XYZ…`) wherever the user is.

The biggest complication would be that we have to copy the python libraries around, which WQ currently does for us.

**Comment by klannon** [comment]
What about starting a local worker and running everything in that worker?

**Comment by matz-e** [comment]
Or do just that… would be nice to not have to dig out a special command for that. Although this also requires a Lobster instance to be running.

I was thinking about investigating why a particular task failed, and that wouldn't be easy if we have to run the master.

**Comment by klannon** [comment]
The more I think about this, the more I think that we absolutely should do this within Lobster rather than rigging a separate way to do it.  I would say we would want to have a `lobster debug` command that more or less does the following:
   1.  Makes a separate WQ master just for running this one task.
   2.  Starts a WQ worker on the local resource to run this task.
   3.  Turns all the debug, verbosity, and monitoring flags up to their maximum levels.  So "debug all" and full resource monitoring instead of just summaries, etc.
   4.  Runs the desired task through this stand-alone machinery.  The screen output should clearly inform the user once Lobster has sent the task off to WQ and when the task has returned, so the user knows when to start doing any interactive monitoring she or he desires.

Does that make sense?  I'll name drop @btovar and @dthain in case they want to offer any suggestions.

Now, for extra credit, I can suggest the following assignments:
   1.  Allow the user to override any parameters of the Lobster task that can be changed at runtime for the purposes of just running this one task.  In particular, I think it could be useful to override the category resource limits to test whether some resource limit is the problem.
   2.  For debug tasks, run Parrot in the mode where it captures all files access and network connections made and include that list to give a more clear picture of what the task is doing in terms of I/O.
   3.  This one is probably the hardest but also particularly interesting to me:  Building on (2) above, but capture the list of files accessed and use this to pack the sandbox sent off to the workers.  Basically, this would be a generic way for Lobster to construct sandboxes for arbitrary tasks.  In other words, if a user can run the task locally, the local worker should be able to do it to (with access to the local files), and Parrot can make a list of what the task actually used and then pack that up.  We'd probably want to implement some filtering so that we don't (automatically anyway) ship things we don't need to (i.e. components of the OS that we expect will be present on the distributed workers or stuff we're accessing "locally" via CVMFS).  Maybe there are smarter ways to accomplish this, but I think that it would be nice for a user to be able to say, "OK, I got this to work locally.  Now, Lobster, just figure out what needs to be packed into the sandbox to run it on my distributed resources."

If that last one sounds like a good idea to you, I can make a separate issue for it.  I think it's linked in that you would build that on top of the pieces you'd put in to debug jobs.

**Comment by matz-e** [comment]
Sounds good (and a lot more complex). For the point in (3): this would be the smartest way to pack the sandbox, as we're only going to need certain bits within `$LOCALRT`, and maybe some external libraries that we currently can't detect. You could just construct a 1 event task to do this…

I don't think we need a separate issue, this sounds like it fits within the realm of what I originally proposed. I think we could accomplish most of this by sub-classing our current task source, and then specify *explicitly* the same resource conditions that the task was run with. This would involve saving said information, still.

**Comment by klannon** [comment]
Can I suggest one tweak to what you describe above?  Instead of saying "construct a 1 event task," I would like to say construct a "1 jobit" or "1 unit" task.  (I feel like our jargon has started to drift here.)  I would really like to be able to get back to the notion that Lobster operates on work units (i.e. jobits) that represent the smallest quantum of work that can sensibly be done.  For data processing, this is probably the lumisection.  For simulation, it's probably something between 10 and 100 events.  In other cases, it might be reading one file.  If we can ingrain this concept into Lobster (i.e. every task has a smallest meaningful quantum), then I think it will be easier to define things like debug mode in an application agnostic way, but I also think it's really useful for working with the auto-task sizing logic and fanciness that we've tried and continue to think about.  Just trying to encourage a consistent way of thinking about this stuff.

**Comment by matz-e** [comment]
I was actually thinking "1 event" to grab the sandbox, to keep computing time low. But I see your points!

**Comment by annawoodard** [comment]
This might not be worthwhile, just brainstorming and throwing it out there. I think we should consider the possibility of building this with an additional debug scenario in mind, where you might want to run an arbitrary task instead of a specific one, before starting your run and submitting your workers. Sometimes when I'm starting a new config from scratch there will be various errors in the configuration, and some of them won't show up until tasks start returning (for example, some kinds of stageout errors). If you have n errors in your config and you are unlucky, it might be n * tasktime minutes before you find all of the errors.   If it adds too much complication then it might not be worth it, but if we can roll it into what @klannon and @matz-e have outlined above so it doesn't require extra pieces, it might be useful.

**Comment by annawoodard** [comment]
(It would need to be slightly different, because we wouldn't be 'replaying' anything as has been suggested above.)
